geomesa-feature/src/main/scala/org/locationtech/geomesa/feature/kryo/SimpleFeatureSerializer.scala

```diff
@@ -116,431 +112,80 @@ class TransformingSimpleFeatureSerializer(sft: SimpleFeatureType, transform: Sim
-      val keyEncoding   = matchEncode(keyClass, null)
-      val valueEncoding = matchEncode(valueClass, null)
-
-      (out: Output, value: AnyRef) => {
-        val map = value.asInstanceOf[JMap[Object, Object]]
-        if (map == null) {
-          out.writeInt(-1): Unit
-        } else {
-          out.writeInt(map.size())
-          map.entrySet.foreach { e => keyEncoding(out, e.getKey); valueEncoding(out, e.getValue) }
-        }
-      }
+    writeAttributes(output, sf)
   }
 
-  /**
-   * Based on the method from geotools WKBWriter. This method is optimized for kryo and simplified from
-   * WKBWriter in the following ways:
-   *
-   * 1. Doesn't save SRID (geomesa didn't use that functionality in WKBWriter)
-   * 2. Doesn't handle dimensions > 2
-   * 3. Doesn't worry about byte order (handled by kryo)
-   * 4. Doesn't use a precision model
-   *
-   * @param out
-   * @param geom
-   */
-  def writeGeometry(out: Output, geom: Geometry): Unit =
-    if (geom == null) {
-      out.write(NULL_BYTE)
-    } else {
-      out.write(NON_NULL_BYTE)
-      geom match {
-        case g: Point =>
-          out.writeInt(WKBConstants.wkbPoint, true)
-          writeCoordinate(out, g.getCoordinateSequence.getCoordinate(0))
-
-        case g: LineString =>
-          out.writeInt(WKBConstants.wkbLineString, true)
-          writeCoordinateSequence(out, g.getCoordinateSequence)
-
-        case g: Polygon => writePolygon(out, g)
-
-        case g: MultiPoint => writeGeometryCollection(out, WKBConstants.wkbMultiPoint, g)
-
-        case g: MultiLineString => writeGeometryCollection(out, WKBConstants.wkbMultiLineString, g)
-
-        case g: MultiPolygon => writeGeometryCollection(out, WKBConstants.wkbMultiPolygon, g)
-
-        case g: GeometryCollection => writeGeometryCollection(out, WKBConstants.wkbGeometryCollection, g)
-      }
-    }
+  def writeWithUserData(kryo: Kryo, output: Output, sf: SimpleFeature) = {
+    defaultWrite(kryo, output, sf)
 
-  def writePolygon(out: Output, g: Polygon): Unit = {
-    out.writeInt(WKBConstants.wkbPolygon, true)
-    writeCoordinateSequence(out, g.getExteriorRing.getCoordinateSequence)
-    out.writeInt(g.getNumInteriorRing, true)
-    var i = 0
-    while (i < g.getNumInteriorRing) {
-      writeCoordinateSequence(out, g.getInteriorRingN(i).getCoordinateSequence)
-      i += 1
-    }
+    val kw = KryoSimpleFeatureEncodingsCache.getAbstractWriter
+    kw.writeGenericMap(output, sf.getUserData)
   }
 
-  def writeGeometryCollection(out: Output, typ: Int, g: GeometryCollection): Unit = {
-    out.writeInt(typ, true)
-    out.writeInt(g.getNumGeometries, true)
-    var i = 0
-    while (i < g.getNumGeometries) {
-      writeGeometry(out, g.getGeometryN(i))
-      i += 1
-    }
-  }
+  def defaultRead(kryo: Kryo, input: Input, typ: Class[SimpleFeature]): SimpleFeature = {
+    // read and store version
+    val version = input.readInt(true)
+    KryoSimpleFeatureDecodingsCache.getAbstractReader.version = version
```

Comment:
default read is pretty simple... maybe just duplicate the code in the readWithVis method.